nova/virt/vmwareapi/vmops.py

```diff
@@ -2253,46 +2406,161 @@ def _relocate_vm(self, vm_ref, context, instance, network_info,
+
+        If the destination is another vCenter service, it creates a new
+        VMwareAPISession to the destination vCenter service to gather the
+        needed info for cloning the vm: datastore, service locator, folder.
+        """
+        image_meta = instance.image_meta
+        storage_policy = self._get_storage_policy(instance.flavor)
+        allowed_ds_types = ds_util.get_allowed_datastore_types(
+            image_meta.properties.hw_disk_type)
+
+        # Determine whether we should connect to the dest vCenter endpoint
+        # (if the instance is to be migrated outside of current vCenter)
+        remote = self._is_remote(dest)
+
+        service_spec = None
+        if remote:
+            session = self._create_dest_session(dest)
+            # set the instance_uuid needed in _get_service_locator_spec
+            dest.update({
+                'instance_uuid':
+                    session.vim.service_content.about.instance_uuid})
+            service_spec = self._get_service_locator_spec(dest)
+            cluster = vutil.get_moref(dest['cluster'],
+                                      'ClusterComputeResource')
+            datastore_regex = dest['datastore_regex']
+            # For migration to other vCenter service the disk_move_type
+            # needs to be moveAllDiskBackingsAndDisallowSharing
+            disk_move_type = "moveAllDiskBackingsAndDisallowSharing"
+        else:
+            session = self._session
+            cluster = self._cluster
+            datastore_regex = self._datastore_regex
+            disk_move_type = "moveAllDiskBackingsAndAllowSharing"
+
+        res_pool = vm_util.get_res_pool_ref(session, cluster)
+        datastore = ds_util.get_datastore(session, cluster,
+                                          datastore_regex,
+                                          storage_policy,
+                                          allowed_ds_types)
+        dc_info = self.get_datacenter_ref_and_name(datastore.ref)
+        folder = self._get_project_folder(dc_info, instance.project_id,
+                                          'Instances', session=session)
+
+        rel_spec = vm_util.relocate_vm_spec(self._session.vim.client.factory,
+                                            disk_move_type=disk_move_type,
+                                            res_pool=res_pool, folder=folder,
+                                            datastore=datastore.ref,
+                                            service=service_spec)
+        cloned_vm = self._clone_vm(vm_ref, rel_spec, folder,
+                                   name=self._get_migration_vm_name(instance))
+        # keep track of the cloned vm by assigning migration.uuid to it
+        migration = objects.Migration.get_by_id_and_instance(
+            context, instance.migration_context.migration_id, instance.uuid)
+        vm_util.change_vm_instance_uuid(session, cloned_vm, migration.uuid)
```

Comment:
But the original instance will still be in an incosistent state. Like I detailed in the previous comment, if we catch an exception, what shall we do with it? Should we implement a complete rollback instead of letting nova to ERROR the instance?

Reply:
as in the previous comment: yes. we should make the instance as usable as possible again, because the alternative is getting called in the middle of the weekend/night, because L2 support also can't fix that easily.

Reply:
I've added a catch for this, that would delete the clone in case of failure, and reraise the exception for the upper handling.

Reply:
Don't we have to rollback then, too?